compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,320 @@
+                    )) if place.local == discr => {
+                        if finder.mutated {
+                            debug!(
+                                "The discriminant: {discr:?} was mutated in predecessor: {pred:?}"
+                            );
+                            // We can't optimize this predecessor, so try the next one.
+                            finder.mutated = false;
+
+                            continue 'preds;
+                        }
+
+                        let discr_ty = body.local_decls[discr].ty;
+                        if let Some(discr) = discr_ty.discriminant_for_variant(tcx, variant) {
+                            debug!(
+                                "{pred:?}: {place:?} = {discr_ty:?}::{variant:?}; goto -> {block:?}",
+                            );
+
+                            let target = targets.target_for_value(discr.val);
+                            static_switches
+                                .entry(block)
+                                .and_modify(|static_switches: &mut &mut [StaticSwitch]| {
+                                    if static_switches.iter_mut().all(|switch| {
+                                        if switch.pred == pred {
+                                            switch.target = target;
+                                            false
+                                        } else {
+                                            true
+                                        }
+                                    }) {
+                                        *static_switches =
+                                            tcx.arena.alloc_from_iter(
+                                                static_switches.iter().copied().chain([
+                                                    StaticSwitch { pred, target, exclude },
+                                                ]),
+                                            );
+                                    }
+                                })
+                                .or_insert_with(|| {
+                                    tcx.arena.alloc([StaticSwitch { pred, target, exclude }])
+                                });
+                        }
+
+                        continue 'preds;
+                    }
+                    _ if finder.mutated => {
+                        debug!("The discriminant: {discr:?} was mutated in predecessor: {pred:?}");
+                        // Note that the discriminant could have been mutated in one predecessor
+                        // but not the others, in which case only the predecessors which did not mutate
+                        // the discriminant can be optimized.
+                        finder.mutated = false;
+
+                        continue 'preds;
+                    }
+                    _ => finder.visit_statement(statement, Location { block, statement_index }),
```

Comment:
Could we create `finder`, visit, and check `mutated` in this arm, instead of the other arms?